src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,506 @@
+
+    // temporary debugging aid so that its obvious the class is being used without being spam
+    private static boolean LOG_FIRST_ACCESS = false;
+
+    public CompactHashStore() {
+        this(DEFAULT_INITIAL_CAPACITY);
+    }
+
+    // private non-allocating constructor for .copy(), not part of the interface
+    private CompactHashStore(int[] index, Object[] kvs, int kvStoreInsertionPos, int numSlotsForIndexRebuild) {
+        this.index = index;
+        this.kvStore = kvs;
+        this.kvStoreInsertionPos = kvStoreInsertionPos;
+        this.numSlotsForIndexRebuild = numSlotsForIndexRebuild;
+    }
+
+    public CompactHashStore(int capacity) {
+        // temporary and ugly
+        if (LOG_FIRST_ACCESS) {
+            System.out.println("~~~~~~~~~~~~~~~~~~~~~~~ Using Compact Hashes from Hash Tables That Don't Hate You.");
+            LOG_FIRST_ACCESS = false; // without this we would spam the output stream
+        }
+        assertConstructorPreconditions(capacity);
+
+        int kvCapacity = roundUpwardsToNearestPowerOf2(capacity);
+        // the index array needs to be a little bit sparse for good performance (i.e. low load factor)
+        // so to store 8 key-value entries an index of exactly 8 slots is too crowded
+        int indexCapacity = 2 * kvCapacity;
+
+        // All 0s by default, so all slots are marked empty for free
+        this.index = new int[2 * indexCapacity];
+        this.kvStore = new Object[2 * kvCapacity];
+        this.kvStoreInsertionPos = 0; // always increases by 2, never decreases
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
+    }
+
+    private static int roundUpwardsToNearestPowerOf2(int num) {
+        return Integer.highestOneBit(num - 1) << 1;
+    }
+
+    private static void assertConstructorPreconditions(int capacity) {
+        if (capacity < 1 || capacity > (1 << 28)) {
+            throw CompilerDirectives.shouldNotReachHere();
+        }
+    }
+
+    @ExportMessage
+    protected boolean set(RubyHash hash, Object key, Object value, boolean byIdentity,
+            @Cached @Shared HashingNodes.ToHash hashFunction,
+            @Cached @Shared GetHashPosInIndexNode getHashPos,
+            @Cached @Shared GetKeyPosInKVStoreNode getKey,
+            @Cached FreezeHashKeyIfNeededNode freezeKey,
+            @Cached @Exclusive InlinedConditionProfile keyNotFound,
+            @Bind("$node") Node __) { /* boilerplate argument for InlinedConditionProfile */
+        Object frozenKey = freezeKey.executeFreezeIfNeeded(key, byIdentity);
+        int keyHash = hashFunction.execute(frozenKey, byIdentity);
+        int hashPos = getHashPos.execute(keyHash, index);
+        int keyKvPos = getKey.execute(frozenKey, keyHash, hashPos, byIdentity, index, kvStore);
+
+        // key doesn't exist, so set is a relatively expensive insertion
+        if (keyNotFound.profile(__, keyKvPos == -1)) {
```

Comment:
Good idea, sorry about that.

As you anticipated in a following comment, that -1 actually doesn't mean "deleted", it just means "not found", possibly because it was never inserted in the first place. 

I didn't think of adding a constant for it because I'm too used to array-search code returning -1 for "not found", but it never hurts to name things.